Validity Window Syncer

viquezclaudio opened the pull request:
The Validity Window Syncer is a component that operates between the macro sync and the live sync
Its purpose is to sync all the historic transactions that happened during the current validity window
This is because, as a validator, we need to enforce the "tx in validity window" check when producing blocks

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

viquezclaudio committed:
Initial validity window syncer

Initial validity window synchronization functionality

viquezclaudio committed:
Move the validity syncer into light macro sync

viquezclaudio committed:
Use the Sync Queue

viquezclaudio committed:
Modify the validity start point

Sync the full epoch if the validity window is located within the current epoch.
This will guarantee that once the validity syncing is complete,
the history root can also be verified
Use the transaction proofs
Use the right block when generating validity start proofs
Fix usage of verifier block numbers for history chunks
Various fixes/improvements

viquezclaudio committed:
Revisit can_enforce_validity_window

Modify the conditions that are used to enforce the validity window
Minor cleanup and refactoring
Remove the can_verify_history flag
Sync full epochs
Remove the validity start proof and process

jsdanielh commented on blockchain/src/blockchain/wrappers.rs:
```diff
@@ -258,34 +258,27 @@ impl Blockchain {
-    /// - can enforce transaction validity window
-    /// The second subsumes the first.
-    /// All conditions are checked against the next block (which potentially hasn't been produced yet).
+    /// Check if we have enough state to check for duplicates during the validity window
```
```suggestion
    /// Check if we have enough state to check for duplicate transactions during the validity window
```

jsdanielh commented on blockchain/tests/blockchain.rs:
```diff
@@ -14,6 +14,7 @@ use nimiq_test_utils::{
 };
 
 #[test(tokio::test)]
+#[ignore]
```
Add a `TODO`

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -19,7 +19,7 @@ use thiserror::Error;
 use crate::error::SubscribeToAddressesError;
 
-mod handlers;
+pub mod handlers;
```
`pub(crate)` maybe?

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -49,6 +49,8 @@ pub const MAX_REQUEST_TRANSACTIONS_BY_ADDRESS: u32 = 1000;
 pub const MAX_REQUEST_TRIE_PROOF: u32 = 1000;
 /// The max number of Block proof requests per peer.
 pub const MAX_REQUEST_BLOCKS_PROOF: u32 = 1000;
+/// The max number of ValidityWindow requests per peer.
```
```suggestion
/// The max number of validity window start requests per peer.
```

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -500,3 +502,29 @@ impl Topic for AddressSubscriptionTopic {
+/// The server will provide to us the validity window start proof
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestValidityWindowStart {
+    pub macro_head_number: u32,
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -500,3 +502,29 @@ impl Topic for AddressSubscriptionTopic {
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestValidityWindowStart {
+    pub macro_head_number: u32,
+    pub macro_head_hash: Blake2bHash,
```
Could you add some `rustdoc`?

viquezclaudio committed:
History root as validity sync completion

This fixes #1473
- Use the history root as the condition to determine if we finished doing the validity sync.
- Simplify some logic
- Apply PR comments

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -49,6 +49,8 @@ pub const MAX_REQUEST_TRANSACTIONS_BY_ADDRESS: u32 = 1000;
 /// The max number of Block proof requests per peer.
 pub const MAX_REQUEST_BLOCKS_PROOF: u32 = 1000;
+/// The max number of ValidityWindow requests per peer.
+pub const MAX_REQUEST_VALIDITY_WINDOW_START: u32 = 1000;
```
Do we still need it?

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+/// However, if the validity start is located in the current epoch
+/// we will move it to the beggining of the current epoch,
+/// in order to sync the full epoch and being able to verify the history root.
+pub fn compute_validity_start(macro_head: u32) -> u32 {
```
This function logic I thin could be simpler

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+        );
+    }
+
+    pub fn poll_validity_window_chunks(
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+        while let Poll::Ready(Some(Ok((request, result, peer_id)))) =
+            self.validity_queue.poll_next_unpin(cx)
+        {
+            log::trace!(peer=%peer_id, chunk_index= request.chunk_index, block_number =request.block_number,  "Processing response from validity queue");
```
```suggestion
            log::trace!(peer=%peer_id, chunk_index=request.chunk_index, block_number=request.block_number,  "Processing response from validity queue");
```

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+                            // We need to check if there was an election in between, if so, we need to proceed to the next epoch
+                            if peer_request.election_in_window {
+                                log::trace!(
+                                    "Moving to the next epoch to continue syncing, current epoch {} ",
+                                    Policy::epoch_at(verifier_block_number)
```
```suggestion
                                    current_epoch = Policy::epoch_at(verifier_block_number),
                                    "Moving to the next epoch to continue syncing",
                                    
```

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+
+                        self.validity_queue.add_ids(vec![(request, None)]);
+                    } else {
+                        // If the chunk doesnt verify we disconnect from the peer
```
```suggestion
                        // If the chunk doesn't verify we disconnect from the peer
```

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,316 @@
+                                    verifier_block=request.block_number,
+                                    epoch=request.epoch_number,
+                                    "The validity history chunk didn't verify, disconnecting from peer");
+
```
I would even ban the peer

viquezclaudio committed:
Move blockchain upgradable lock acquisition

Within the extend and rebranch functios, the blockchain lock was upgraded after the
database commit call, which caused an inconsistency between the blockchain and the database state,
however, the previous approach was faster, since the write lock was not held during the entire function execution.

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,304 @@
+use std::{
+    cmp,
```
```suggestion
```

viquezclaudio replied:
done

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -0,0 +1,304 @@
+        let validity_window_bn = if validity_start <= Policy::genesis_block_number() {
+            Policy::genesis_block_number()
+        } else {
+            Policy::election_block_before(validity_start)
```
What if `validity_start` is already an election macro block?

viquezclaudio replied:
we need to sync the previous epoch in order to verify the history

jsdanielh commented:
@viquezclaudio Could you add that this closes #1473?
